### PR TITLE
Target estimator

### PR DESCRIPTION
**Pull request by james-ward**
Uses turret movement to correct exponential moving average estimate of target position.
Also uses chassis movement to correct target position.
Ignores vision information if turret is not pointing downrange.

### CODE REVIEW

File: components/target_estimator.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    ALPHA = 0.1  # Exponential smoothing fraction
+    CAMERA_TO_LIDAR = 0.15
+
+    chassis: Chassis
+    range_finder: RangeFinder
+    turret: Turret
+    vision: Vision
+
+    def setup(self) -> None:
```
Comment (auscompgeek) [inline components/target_estimator.py]:
```suggestion
    def __init__(self) -> None:
```
since this doesn't require any of the injected components.

File: components/target_estimator.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        self.reset()
+
+    def on_enable(self) -> None:
+        self.reset()
+
+    def reset(self) -> None:
+        self.angle_to_target: float = 0.0
+        self.vision_range: float = 0.0
+        self.previous_azimuth: float = 0.0
+        self.previous_heading: float = None
```
Comment (auscompgeek) [inline components/target_estimator.py]:
```suggestion
        self.previous_heading: Optional[float] = None
```
(with the prerequisite `from typing import Optional` of course)

File: robot.py [hunk 3]
```diff
@@ -39,6 +40,7 @@ class MyRobot(magicbot.MagicRobot):
     spinner_controller: SpinnerController
 
     # List components (which represent physical subsystems) here.
+    target_estimator: TargetEstimator
```
Comment (auscompgeek) [inline robot.py]: We probably actually want this to execute before the shoot controller so it can use the latest information, right?

File: components/chassis.py [hunk 4]
```diff
@@ -114,6 +114,10 @@ def get_pose(self) -> Pose2d:
         """Get the current position of the robot on the field."""
         return self.odometry.getPose()
 
+    def get_heading(self) -> float:
+        """Get the current heading of the robot."""
+        return self.get_pose().rotation().radians()
```
Comment (auscompgeek) [inline components/chassis.py]: I'm inclined to have this return the Rotation2d instead. Its addition/subtraction implementation bounds the radians value to [-pi,pi], which looks like it might be useful below.

Comment (james-ward) [reply components/chassis.py]: The reason for this is the amount of boilerplate required to get the heading when you want it. Which tends to be what people want - the upstream API is one that is technically the correct way to do it, and wholly unhelpful for the vast majority of uses!

Comment (james-ward) [reply components/chassis.py]: Also, I can change to doing all the differences with `Pose2d` or `Rotation2d` objects, but I don't think it will reduce the number of calls to a constraining function - the geometry library must be making that call every time a difference is calculated any way.
tl;dr - I'll change this if the consensus is that it's a better way to do it.

Comment (auscompgeek) [reply components/chassis.py]: Might not be worth it unless we use it consistently. Maybe let's see how we go. If we think this has become too messy we can see how swapping everything over goes.

File: components/target_estimator.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        delta = constrain_angle(current_heading - self.previous_heading)
+        self.angle_to_target -= delta
+        self.previous_azimuth = current_azimuth
+        self.previous_heading = current_heading
+
+        # Check for new vision data
+        # We only add it if it could possibly be a target
+        # Calculate the turret azimuth in field coordinates
+        turret_in_field = constrain_angle(current_heading + current_azimuth)
+        self._pointing_downrange = abs(turret_in_field) < math.pi / 2.0
```
Comment (raulvera59) [inline components/target_estimator.py]: As it stands this relies on the turret azimuth **not** being in robot coordinates. (As far as I can tell, field coordinates are whatever the robot coordinates were when the imu was powered up, i.e. a heading of 0 is pointing up range.) Once the PR to fix that is merged, this will break. The correct test will be turret_in_field < -pi/2 or turret_in_field > pi/2.

File: components/target_estimator.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    def get_data(self) -> VisionData:
+        if abs(self.angle_to_target) < math.radians(5.0):
+            vd = VisionData(
+                self.range_finder.get_distance() - self.CAMERA_TO_LIDAR,
+                self.angle_to_target,
+                time.monotonic(),
+            )
+        else:
+            vd = VisionData(self.vision_range, self.angle_to_target, time.monotonic(),)
```
Comment (raulvera59) [inline components/target_estimator.py]: If we can't see a target, this returns bad data. See my PR comment for an alternative.

File: components/target_estimator.py (hunk 5 above)
Comment (james-ward) [reply components/target_estimator.py]: Field coordinates are fixed at 0 degrees pointing downfield away from our alliance station. So I think that means this is actually correct.

File: components/target_estimator.py (hunk 6 above)
Comment (james-ward) [reply components/target_estimator.py]: I'm just worried about turning this into a full blown state estimator. Once we start fusing odometry it gets really complicated and I think we'd be better off using a proper ukf to do it. That means tuning a ukf...
I do want to be able to maintain the estimate when we lose sight during an unwind of the turret - this currently just resets the filter and starts again. I can't see a nice way to do that until we have the turret logic moved out into your turret controller.

Comment (james-ward) [reply components/target_estimator.py]: I'm just worried about turning this into a full blown state estimator. Once we start fusing odometry it gets really complicated and I think we'd be better off using a proper ukf to do it. That means tuning a ukf...
I do want to be able to maintain the estimate when we lose sight during an unwind of the turret - this currently just resets the filter and starts again. I can't see a nice way to do that until we have the turret logic moved out into your turret controller.

Comment (raulvera59) [reply components/target_estimator.py]: I don't follow. If we know our position and heading on the field and the azimuth of our turret, we can estimate the azimuth of the target because we know where it is on the field. It isn't fusing odometry any more than you are already doing. It doesn't need to be particularly accurate, just enough to get us back to where we can see the target. The current plan is to merge my turret changes, then pull this PR and make these tweaks.

File: components/target_estimator.py (hunk 5 above)
Comment (raulvera59) [reply components/target_estimator.py]: So the robot heading starts off at 0 toward the back of the robot, matching the field. Ok. So the turret coordinate system starts off aligned with the field. Cool.

Comment (raulvera59) [reply components/target_estimator.py]: And the current azimuth is in robot coordinates (or will be), so the test might not need to change but the computation will, because the field is 180 degrees from robot coordinates.

Comment (raulvera59) [reply components/target_estimator.py]: And once the turret get_azimuth is fixed, we'll have to convert it to the field coordinates, so the computation will change.

Comment (auscompgeek) [reply components/target_estimator.py]: I was under the impression we weren't assuming we were going to have the robot heading start at 0.

Comment (raulvera59) [reply components/target_estimator.py]: This code seems to assume that.

Comment (james-ward) [reply components/target_estimator.py]: The autonomous routine knows the heading of the robot and initialises the pose accordingly.

File: components/turret.py [hunk 7]
```diff
@@ -288,7 +288,11 @@ def azimuth_at_time(self, t: float) -> Optional[int]:
         current_time = time.monotonic()
         control_loops_ago = int((current_time - t) / self.control_loop_wait_time)
         if control_loops_ago >= len(self.azimuth_history):
-            return None
```
Comment (mlists) [inline components/turret.py]: The typehint for this function no longer needs to be optional

File: controllers/shooter.py [hunk 8]
```diff
@@ -147,6 +135,7 @@ def find_allowable_angle(self, dist: float) -> float:
         Currently does not consider angle from target
         dist: planar distance from the target
         """
+        dist = min(dist, 14.0)
```
Comment (mlists) [inline controllers/shooter.py]: Should we consider clamping with a max as well?

File: components/target_estimator.py (hunk 6 above)
Comment (james-ward) [reply components/target_estimator.py]: We don't know our position on the field very well. We initialise the chassis with something close to the truth, given the accuracy of the drive team placing the robot. We then dead reckon our way around the field and the error accumulates, and will get especially bad when we drive over the metal in the middle of the field. I think using odometry without any way to correct our absolute position will cause problems.

File: controllers/shooter.py (hunk 8 above)
Comment (james-ward) [reply controllers/shooter.py]: Or would be 0 surely? Not a bad idea.

File: components/target_estimator.py (hunk 6 above)
Comment (raulvera59) [reply components/target_estimator.py]: Ok. I've just changed the scan start point to be just plain downrange, i.e. the negative of the heading. That's very coarse but no worse than the corrections the target estimator already does, and it should be good enough to start scanning in the right general vicinity.

**Comment by raulvera59** [comment]
I've merged this PR into a local branch and am setting up a PR including this and my subsequent changes. I've changed essentially nothing of James' code beyond fixing conflicts with the latest turret code.
